c2rust/src/main.rs

```diff
@@ -1,56 +1,154 @@
 use git_testament::{git_testament, render_testament};
+use is_executable::IsExecutable;
+use std::borrow::Cow;
+use std::collections::HashMap;
 use std::env;
 use std::ffi::OsStr;
-use std::process::{exit, Command};
+use std::path::{Path, PathBuf};
+use std::process;
+use std::process::Command;
+use std::str;
 
 git_testament!(TESTAMENT);
 
-fn main() {
-    let subcommand_yamls = [load_yaml!("transpile.yaml")];
-    let matches = App::new("C2Rust")
-        .version(&*render_testament!(TESTAMENT))
-        .author(crate_authors!(", "))
-        .setting(AppSettings::SubcommandRequiredElseHelp)
-        .subcommands(
-            subcommand_yamls
-                .iter()
-                .map(|yaml| SubCommand::from_yaml(yaml)),
-        )
-        .get_matches();
+/// A `c2rust` sub-command.
+struct SubCommand {
+    /// The path to the [`SubCommand`]'s executable,
+    /// if it was found (see [`Self::find_all`]).
+    /// Otherwise [`None`] if it is a known [`SubCommand`] (see [`Self::known`]).
+    path: Option<PathBuf>,
+    /// The name of the [`SubCommand`], i.e. in `c2rust-{name}`.
+    name: Cow<'static, str>,
+}
 
-    let mut os_args = env::args_os();
-    os_args.next(); // Skip executable name
-    let arg_name = os_args.next().and_then(|name| name.into_string().ok());
-    match (&arg_name, matches.subcommand_name()) {
-        (Some(arg_name), Some(subcommand)) if arg_name == subcommand => {
-            invoke_subcommand(subcommand, os_args);
-        }
-        _ => {
-            eprintln!("{:?}", arg_name);
-            panic!("Could not match subcommand");
+impl SubCommand {
+    /// Find all [`SubCommand`]s adjacent to the current (`c2rust`) executable.
+    /// They are of the form `c2rust-{name}`.
+    pub fn find_all() -> anyhow::Result<Vec<Self>> {
+        let c2rust = env::current_exe()?;
+        let c2rust_name: &Utf8Path = c2rust
+            .file_name()
+            .map(Path::new)
+            .ok_or_else(|| anyhow!("no file name: {}", c2rust.display()))?
+            .try_into()?;
+        let c2rust_name = c2rust_name.as_str();
+        let dir = c2rust
+            .parent()
+            .ok_or_else(|| anyhow!("no directory: {}", c2rust.display()))?;
+        let mut sub_commands = Vec::new();
+        for entry in dir.read_dir()? {
+            let entry = entry?;
+            let file_type = entry.file_type()?;
+            if !(file_type.is_file() || file_type.is_symlink()) {
+                continue;
+            }
+            let file_name = entry.file_name();
+            let file_name: &Utf8Path = Path::new(&file_name).try_into()?;
+            let file_name = file_name.as_str();
+            let name = match file_name.strip_prefix(c2rust_name) {
+                Some(name) => name,
+                None => continue,
+            };
+            let name = match name.strip_prefix('-') {
+                Some(name) => name,
+                None => continue,
+            };
```

Comment:
Rather than doing two strips, I think it would be clearer to append `-` to the name and strip that prefix once. Maybe a bit less efficient in memory, but we don't care here.

Reply:
How about this instead?  https://github.com/immunant/c2rust/blob/5c4867130b5430ee7a343965e07c1375af48e59e/c2rust/src/main.rs#L42-L60